packages/flutter/lib/src/material/tabs.dart

```diff
@@ -432,100 +469,132 @@ class TabBarSelection {
     _performance
       ..progress = progress
       ..forward().then((_) {
-        if (onChanged != null)
-          onChanged();
-        _indexIsChanging = false;
+        if (_performance.progress == 1.0) {
+          if (config.onChanged != null)
+            config.onChanged(_index);
+          _indexIsChanging = false;
+        }
       });
   }
 
   int get previousIndex => _previousIndex;
   int _previousIndex = 0;
+
+  final List<TabBarSelectionPerformanceListener> _performanceListeners = <TabBarSelectionPerformanceListener>[];
+
+  void registerPerformanceListener(TabBarSelectionPerformanceListener listener) {
+    _performanceListeners.add(listener);
+    _performance
+      ..addStatusListener(listener.handleStatusChange)
+      ..addListener(listener.handleProgressChange);
+  }
+
+  void unregisterPerformanceListener(TabBarSelectionPerformanceListener listener) {
+    _performanceListeners.remove(listener);
+    _performance
+      ..removeStatusListener(listener.handleStatusChange)
+      ..removeListener(listener.handleProgressChange);
+  }
+
+  void deactivate() {
+    for (TabBarSelectionPerformanceListener listener in _performanceListeners.toList()) {
+      listener.handleDeactivate();
+      unregisterPerformanceListener(listener);
+    }
+    assert(_performanceListeners.isEmpty);
+  }
+
```

Comment:
I bet we don't handle someone _inserting_ a `TabBarSelection` above a `TabBar` (but in a way that doesn't cause the `TabBar` to build) correctly. No idea how to make that work...